File: frontend/src/renderer/services/TargetService.ts
/**
 * TargetService - 目标管理服务
 */

import { getDatabase, CreateTargetRequest, UpdateTargetRequest, Target } from './database/DatabaseService';

export class TargetService {
  private db = getDatabase();

  /**
   * 获取所有目标
   */
  getTargets(): Target[] {
    return this.db.getTargets();
  }

  /**
   * 获取单个目标
   */
  getTarget(id: number): Target | null {
    return this.db.getTarget(id);
  }

  /**
   * 创建目标
   */
  createTarget(data: CreateTargetRequest): Target {
    // 验证 URL
    if (!this.validateUrl(data.url)) {
      throw new Error('Invalid URL format');
    }

    // 检查是否已存在
    const existing = this.db.getTargets().find(t => t.url === data.url);
    if (existing) {
      throw new Error('Target with this URL already exists');
    }

    const id = this.db.createTarget(data);
    return this.db.getTarget(id)!;
  }

  /**
   * 更新目标
   */
  updateTarget(id: number, data: UpdateTargetRequest): Target {
    if (!this.db.getTarget(id)) {
      throw new Error('Target not found');
    }

    // 如果更新 URL，验证新 URL
    if (data.url && !this.validateUrl(data.url)) {
      throw new Error('Invalid URL format');
    }

    this.db.updateTarget(id, data);
    return this.db.getTarget(id)!;
  }

  /**
   * 删除目标
   */
  deleteTarget(id: number): void {
    if (!this.db.getTarget(id)) {
      throw new Error('Target not found');
    }

    this.db.deleteTarget(id);
  }

  /**
   * 验证 URL 格式
   */
  validateUrl(url: string): boolean {
    try {
      const parsed = new URL(url);
      return ['http:', 'https:'].includes(parsed.protocol);
    } catch {
      return false;
    }
  }

  /**
   * 搜索目标
   */
  searchTargets(query: string): Target[] {
    const targets = this.db.getTargets();
    const lowerQuery = query.toLowerCase();

    return targets.filter(
      t =>
        t.name.toLowerCase().includes(lowerQuery) ||
        t.url.toLowerCase().includes(lowerQuery) ||
        t.tags.some(tag => tag.toLowerCase().includes(lowerQuery))
    );
  }
}

export default new TargetService();

File: backend/internal/services/target_service.go
package services

import (
	"database/sql"
	"encoding/json"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/holehunter/backend/internal/models"
)

type TargetService struct {
	db *sql.DB
}

func NewTargetService(db *sql.DB) *TargetService {
	return &TargetService{db: db}
}

func (s *TargetService) ListTargets(c *gin.Context) {
	rows, err := s.db.Query("SELECT id, name, url, description, tags, created_at, updated_at FROM targets ORDER BY created_at DESC")
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer rows.Close()

	targets := []models.Target{}
	for rows.Next() {
		var t models.Target
		var tagsStr string
		if err := rows.Scan(&t.ID, &t.Name, &t.URL, &t.DESC, &tagsStr, &t.CreatedAt, &t.UpdatedAt); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		t.ScanTags(tagsStr)
		targets = append(targets, t)
	}

	c.JSON(http.StatusOK, gin.H{"data": targets})
}

func (s *TargetService) GetTarget(c *gin.Context) {
	id := c.Param("id")
	var t models.Target
	var tagsStr string

	err := s.db.QueryRow(
		"SELECT id, name, url, description, tags, created_at, updated_at FROM targets WHERE id = ?",
		id,
	).Scan(&t.ID, &t.Name, &t.URL, &t.DESC, &tagsStr, &t.CreatedAt, &t.UpdatedAt)

	if err == sql.ErrNoRows {
		c.JSON(http.StatusNotFound, gin.H{"error": "Target not found"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	t.ScanTags(tagsStr)
	c.JSON(http.StatusOK, gin.H{"data": t})
}

func (s *TargetService) CreateTarget(c *gin.Context) {
	var req models.CreateTargetRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	tagsJSON, _ := json.Marshal(req.Tags)
	result, err := s.db.Exec(
		"INSERT INTO targets (name, url, description, tags) VALUES (?, ?, ?, ?)",
		req.Name, req.URL, req.DESC, string(tagsJSON),
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	id, _ := result.LastInsertId()
	c.JSON(http.StatusCreated, gin.H{
		"success": true,
		"data": gin.H{"id": id},
	})
}

func (s *TargetService) UpdateTarget(c *gin.Context) {
	id := c.Param("id")
	var req models.CreateTargetRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	tagsJSON, _ := json.Marshal(req.Tags)
	_, err := s.db.Exec(
		"UPDATE targets SET name = ?, url = ?, description = ?, tags = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?",
		req.Name, req.URL, req.DESC, string(tagsJSON), id,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

func (s *TargetService) DeleteTarget(c *gin.Context) {
	id := c.Param("id")

	_, err := s.db.Exec("DELETE FROM targets WHERE id = ?", id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}
